hash_writer_test.go

```diff
@@ -0,0 +1,340 @@
+		t.Errorf("expected ionHashWriter.stepOut() to execute without errors; %s", err.Error())
+	}
+
+	sum, err = ionHashWriter.Sum(nil)
+	if err != nil {
+		t.Errorf("expected Sum(nil) to execute without errors; %s", err.Error())
+	}
+
+	expectedSum = []byte{0x0b, 0xb0, 0x0b, 0x20, 0x05, 0x0e, 0x0e}
+
+	if !reflect.DeepEqual(sum, expectedSum) {
+		t.Errorf("expected sum to be %v instead of %v", expectedSum, sum)
+	}
+
+	err = ionHashWriter.WriteNull()
+	if err != nil {
+		t.Errorf("expected ionHashWriter.WriteNull() to execute without errors; %s", err.Error())
+	}
+
+	sum, err = ionHashWriter.Sum(nil)
+	if err != nil {
+		t.Errorf("expected Sum(nil) to execute without errors; %s", err.Error())
+	}
+
+	expectedSum = []byte{0x0b, 0x0f, 0x0e}
+
+	if !reflect.DeepEqual(sum, expectedSum) {
+		t.Errorf("expected sum to be %v instead of %v", expectedSum, sum)
+	}
+
+	if ionHashWriter.isInStruct() {
+		t.Error("expected ionHashWriter.isInStruct() to return false")
+	}
+
+	err = ionHashWriter.stepIn(ion.StructType)
+	if err != nil {
+		t.Errorf("expected ionHashWriter.stepIn(ion.StructType) to execute without errors; %s", err.Error())
+	}
+
+	if !ionHashWriter.isInStruct() {
+		t.Error("expected ionHashWriter.isInStruct() to return true")
+	}
+
+	ionHashWriter.setFieldName("hello")
+
+	err = ionHashWriter.Annotation("ion")
+	if err != nil {
+		t.Errorf("expected ionHashWriter.Annotation(\"ion\") to execute without errors; %s", err.Error())
+	}
+
+	err = ionHashWriter.Annotation("hash")
+	if err != nil {
+		t.Errorf("expected ionHashWriter.Annotation(\"hash\") to execute without errors; %s", err.Error())
+	}
+
+	err = ionHashWriter.WriteSymbol("world")
+	if err != nil {
+		t.Errorf("expected ionHashWriter.WriteSymbol(\"world\") to execute without errors; %s", err.Error())
+	}
+
+	err = ionHashWriter.stepOut()
+	if err != nil {
+		t.Errorf("expected ionHashWriter.stepOut() to execute without errors; %s", err.Error())
+	}
+
+	if ionHashWriter.isInStruct() {
+		t.Error("expected ionHashWriter.isInStruct() to return false")
+	}
+
+	sum, err = ionHashWriter.Sum(nil)
+	if err != nil {
+		t.Errorf("expected Sum(nil) to execute without errors; %s", err.Error())
+	}
+
+	expectedSum = []byte{0x0b, 0xd0,
+		0x0c, 0x0b, 0x70, 0x68, 0x65, 0x6c, 0x6c, 0x6f, 0x0c, 0x0e, // hello:
+		0x0c, 0x0b, 0xe0,
+		0x0c, 0x0b, 0x70, 0x69, 0x6f, 0x6e, 0x0c, 0x0e, // ion::
+		0x0c, 0x0b, 0x70, 0x68, 0x61, 0x73, 0x68, 0x0c, 0x0e, // hash::
+		0x0c, 0x0b, 0x70, 0x77, 0x6f, 0x72, 0x6c, 0x64, 0x0c, 0x0e, // world
+		0x0c, 0x0e,
+		0x0e}
+
+	if !reflect.DeepEqual(sum, expectedSum) {
+		t.Errorf("expected sum to be %v instead of %v", expectedSum, sum)
+	}
+
+	err = ionHashWriter.Finish()
+	if err != nil {
+		t.Errorf("expected ionHashWriter.Finish() to execute without errors; %s", err.Error())
+	}
+
+	expectedStr := "null [5] null {hello:ion::hash::world}"
+
+	if str.String() != expectedStr {
+		t.Errorf("expected str.String() to return \"%s\" instead of %s", expectedStr, str.String())
+	}
+}
```

Comment:
Good test. If we could easily refactor pieces of it into smaller tests that would be ideal, e.g., 

1. One test that deals with null(s)
1. One that deals with a valid Ion data that are scalars 
1. One that deals with containers (list and structs will do for now)

Reply:
Done.